language/move-binary-format/src/file_format.rs

```diff
@@ -647,23 +664,24 @@ impl AbilitySet {
     }
 
     /// For a polymorphic type, its actual abilities correspond to its declared abilities but
-    /// predicated on its type arguments having that ability. For `Key`, instead of needing
-    /// the same ability, the type arguments need `Store`
+    /// predicated on its non-phantom type arguments having that ability. For `Key`, instead of needing
+    /// the same ability, the type arguments need `Store`.
     pub fn polymorphic_abilities(
         declared_abilities: Self,
-        type_argument_abilities: impl IntoIterator<Item = Self>,
+        type_arguments: impl IntoIterator<Item = (Self, bool)>,
```

Comment:
Sorry for bringing this up so late, but I suppose thats what happens after a weekend and a set of fresh eyes.

I feel like maybe using `ExactSizeIterator` would be more obvious. Originally I said to do the zipped so it was guaranteed. And you will get some weird stuff if they don't line up...  
```
declared_abilities: Self,
declared_phantom_parameters: impl ExactSizeIterator<Item =  bool>,
type_arguments: impl ExactSizeIterator<Item =Self>,
```
but then you'll have to verify inputs which isn't great...

We could also do something like 
```
pub fn polymorphic_abilities<const N: usize>(
    declared_abilities: Self,
    declared_phantom_parameters: &[bool; N],
    type_arguments: &[Self; N]
) -> Self
```
but this might be really hard to setup for a callsite (if not impossible, not sure what utilities exist around arrays off hand)

Reply:
The `ExactSizeIterator` makes sense, but I'm not sure how the function it's supposed to fail in that case.

Another option is to keep this function as it was, and filter phantom parameters at the call site: it's the caller's responsibility to pass in type parameters that are relevant.

Reply:
Maybe just `ExactSizeIterator`, and wrap it in a `PartialVMResult` that gives an invariant violation?

Reply:
That should work in principle, but not sure how much the error has to be propagated up the call stack. It may be the right thing to do though...

I'll put it in a new commit.